### ISSUE TITLE
Regression of #14759: Unable to use Sogou Pinyin to input Chinese characters in NVDA's editable controls

### ISSUE THREAD

**Issue by cary-rowen** [nvaccess/nvda#14784]
### Steps to reproduce:
1. Download and install Sogou Pinyin Input Method:https://shurufa.sogou.com/
2. Find editable control in NVDA, such as filter in input gesture dialogs.
3. ctrl+shift switch input method to Sogou Pinyin.
4. Try typing 'duixiang' and press the space to confirm.

### Actual behavior:
The character actually entered into the edit box is 'duixiang' or blank instead of '对象'.
### Expected behavior:
The Chinese word '对象' is entered into the edit box.

### NVDA logs, crash dumps and other attachments:
[log.txt](https://github.com/nvaccess/nvda/files/11147782/log.txt)

#### NVDA installed/portable/running from source:
Installed
#### NVDA version:
alpha-28022
#### Windows version:
Windows 10 22H2 (AMD64) build 19045.2728
#### Name and version of other software in use when reproducing the issue:
Sogou Pinyin Input Method 13
#### Other information about your system:
None
#### Does the issue still occur after restarting your computer?
Yes
#### Have you tried any other versions of NVDA? If so, please report their behaviors.
The previous alpha version could not reproduce the issue, But the build of #14759 reproduces the issue.
#### If NVDA add-ons are disabled, is your problem still occurring?
Yes
#### Does the issue still occur after you run the COM Registration Fixing Tool in NVDA's tools menu?
Yes

**Comment by cary-rowen** [comment]
cc @leonardder

**Comment by jcsteh** [comment]
Input methods run in-process, so if the input method does call SendMessage, it will indeed be impacted. I don't really understand why it would break though.

I guess we could restrict this to oleacc and UIACore functions by hooking the relevant functions, setting a bool flag and checking that in our SendMessage hooks. The challenge is that we'd have to figure out all the impacted oleacc and UIACore functions. We probably can't even hook some of them, as they may not have symbols; e.g. the various oleacc proxies that send messages to remote processes.

One thing to try is to bypass our hooks if the target HWND is within the current thread. It isn't useful to use cancellableSendMessage in that case anyway, since those calls go direct to the window proc anyway.

**Comment by leonardder** [comment]
> One thing to try is to bypass our hooks if the target HWND is within the current thread. 

I think we already do this but I'll have a look.

**Comment by jcsteh** [comment]
You're right; we do. However, we still mess with the timeout and hung flags, plus check the cancellation event. That shouldn't matter, but I think it'd still be worth seeing if just returning as early as possible makes things better if we're in the same thread.

**Comment by leonardder** [comment]
@cary-rowen It looks like the url in the str is unavailable. I'm getting a 403 error.

**Comment by cary-rowen** [comment]
Sorry I provided a URL that is no longer valid, you can download it via the link below:
https://shurufa.sogou.com/
Once the above page is open press Tab to navigate to the 'Windows' link and activate it to download.

**Comment by cary-rowen** [comment]
I just installed and tested Sogou Pinyin input method in English environment, the software does not have internationalization support, and the installation wizard is in Chinese, which may be a challenge for you.
I provide the following steps so that you can install the software smoothly:
1. Run the installer, there may be a UAC permission, press Alt+Y to allow it.
2. Press tab to navigate to the only one CheckBox to check it.
3. Then press shift +Tab to navigate to a button and press space to click it.
The second step is to check the license agreement, and the third step is to install the software.

**Comment by leonardder** [comment]
Ugh, this feels a bit uncomfortable to me honestly.
Let's perform some additional changes to the pr in the hope that we can track this down otherwise.

**Comment by cary-rowen** [comment]
I agree with you, feel free to experiment with changes and I'll provide testing feedback in a timely manner.

**Comment by leonardder** [comment]
I'm now able to reproduce the issue.
It looks like we can avoid the regression when we don't use the background thread. It looks like that background was introduces due to an issue in PowerPoint 2007, which is no longer in use. So if there are no other benefits introduced by using a background thread, we could consider throwing it away. @jcsteh What do you think?

**Comment by jcsteh** [comment]
I'm a little confused about something. If this message is being sent from the main thread to the main thread, why doesn't the early return happen?

`if (currentThreadId == mainThreadId && GetWindowThreadProcessId(hwnd, NULL) == mainThreadId) {`

That should mean we don't use the background thread.
